File: String/166. Fraction to Recurring Decimal/solution.py
class Solution:
    def fractionToDecimal(self, numerator: int, denominator: int) -> str:
        if numerator%denominator == 0: return str(numerator//denominator)

        diffSign = (numerator > 0) ^ (denominator > 0)
        res = "-" if diffSign else "" # handle sign
        
        A, B = abs(numerator), abs(denominator)
        res += str(A // B) + "."
        
        remainder = A % B
        visited = {}

        while remainder != 0 and remainder not in visited:
            visited[remainder] = len(res)

            remainder *= 10
            res += str(remainder // B)
            remainder = remainder%B
        
        if remainder == 0:
            return res
        else:
            pos = visited[remainder]
            return f"{res[:pos]}({res[pos:]})"

File: String/166. Fraction to Recurring Decimal/solution.cpp
class Solution {
public:
    string fractionToDecimal(int numerator, int denominator) {
        if (numerator == 0) return "0";
        
        string ans = "";
        
        // Handle sign
        if ((numerator > 0) ^ (denominator > 0)) {
            ans += "-";
        }
        
        // Use long long to prevent overflow
        long long dividend = abs((long long)numerator);
        long long divisor = abs((long long)denominator);
        
        // Integer part
        ans += to_string(dividend / divisor);
        dividend %= divisor;
        
        if (dividend == 0) return ans;
        
        // Decimal part
        ans += ".";
        unordered_map<long long, int> seen;
        
        while (dividend != 0) {
            if (seen.count(dividend)) {
                // Found repeating cycle
                int pos = seen[dividend];
                ans.insert(pos, "(");
                ans += ")";
                return ans;
            }
            
            seen[dividend] = ans.size();
            dividend *= 10;
            ans += to_string(dividend / divisor);
            dividend %= divisor;
        }
        
        return ans;
    }
};